### PR TITLE
Feature: Add InfoBox Feature to Marker

### PR DESCRIPTION
**Pull request by devmessias**
Example:

```python
m = gmaps.Map()
locations = [
              (-34.0, -59.166672),
              (-32.23333, -64.433327),
          ]
names = ["Atucha", "Embalse"]
infos_html=["<h1>Atucha</h1>","<h2>Embalse</h2>"]
symbol_layer = gmaps.symbol_layer(locations, hover_text=names,info_html=infos_html)
m = gmaps.Map()
m.add_layer(symbol_layer)
m
```
captura de tela de 2016-12-10 16 08 50

### CODE REVIEW

**Comment by pbugnion** [comment]
This looks great! Thanks for submitting. I'll look at it later today.

File: js/src/jupyter-gmaps.js [hunk 1]
```diff
@@ -183,17 +184,26 @@ export const BaseMarkerView = widgets.WidgetView.extend({
...
+        marker.addListener('click', function() {
+            infoWindow.open(mapView.map, marker);
+        });
     },
 
     modelEvents() {
         // Simple properties:
         const properties = [
-            ['title', 'hover_text']
+            ['title', 'hover_text','info_html']
```
Comment (pbugnion) [inline js/src/jupyter-gmaps.js]: This isn't quite right. It should be:

```javascript
    modelEvents() {
        // Simple properties:
        const properties = [
            ['title', 'hover_text']
        ]
        properties.forEach(([nameInView, nameInModel]) => {
            const callback = (
                () => {
                  this.marker.set(
                  nameInView, this.model.get(nameInModel))
                }
            )
            this.model.on(`change:${nameInModel}`, callback, this)
        })

        const infoBoxProperties = [
            ['content', 'info_html']
            // 'content' is the name of the property for the InfoWindow object,
            // 'info_html' is the name of the property in the Python widget
        ]
        infoBoxProperties.forEach(([nameInView, nameInModel]) => {
            const callback = (
                () => {
                    this.infoWindow.set(
                    nameInView, this.model.get(nameInModel))
                }
            )
            this.model.on(`change:${nameInModel}`, callback, this)
        })

        this.setStyleEvents()
    }
```

The idea is that, every time the property `info_html` is changed at the Python level, this triggers an event in JS called `change:info_html` (by the magic of traitlets and ipywidgets). We need to react to that event by changing the property in the google maps `InfoWindow` object.

This lets us change the info text without having to redraw the entire map:

screen shot 2016-12-11 at 14 19 02

Ultimately, listening to events and handling them correctly should let people build mini-applications running in Jupyter notebooks.

File: js/src/jupyter-gmaps.js [hunk 2]
```diff
@@ -183,17 +184,26 @@ export const BaseMarkerView = widgets.WidgetView.extend({
             ...styleOptions
         }
         this.marker = new google.maps.Marker(markerOptions)
+
+        this.infoWindow=new google.maps.InfoWindow({
```
Comment (pbugnion) [inline js/src/jupyter-gmaps.js]: We normally leave spaces around the `=` sign.

File: gmaps/marker.py [hunk 3]
```diff
@@ -161,22 +165,26 @@ def _symbol_layer_options(
...
         "hover_text": hover_text,
+        "info_html": info_html,
         "label": label
     }
     return _merge_option_dicts(options)
 
 
 def symbol_layer(
-        locations, hover_text="", fill_color=None, fill_opacity=1.0,
+        locations, hover_text="", info_html='', fill_color=None,
```
Comment (pbugnion) [inline gmaps/marker.py]: Could we change the single quotes to double quotes (to be consistent with the rest of the code)?

**Comment by pbugnion** [comment]
Since this adds value, it makes sense to do a minor release as soon as it's ready and appropriately documented. I'll create a milestone.

**Comment by devmessias** [comment]
Hi @pbugnion . First, thanks for your advices. I will implement your requested changes soon as possible.

**Comment by pbugnion** [comment]
This looks good! Thanks for fixing these up.

Are you happy for me to merge this?

**Comment by pbugnion** [comment]
Merging into infobox branch (not master) to fix regression with info box showing up even when the text content is empty.

**Comment by devmessias** [comment]
Sorry, I was on vacation. I had implemented something to prevent empty information boxes but I do not know why I'd not send a pull request

```javascript
export const BaseMarkerView = widgets.WidgetView.extend({
    render() {
 //....
    this.infoWindow = infoBoxHml != '' ? new google.maps.InfoWindow({
            content: infoBoxHml
        }):null;
        // ....
    },

    addToMapView(mapView) {
  //...
        let infoWindow = this.infoWindow;
        
        if (infoWindow != null){
        marker.addListener('click', function() {
            infoWindow.open(mapView.map, marker);
        });
}
```

maybe, because this solution is ugly.

..have you fixed this bug?

**Comment by pbugnion** [comment]
No worries! The implementation I went for is pretty similar to your suggestion, with two differences:

 - I include an extra argument called `display_info_box` when building the model. This defaults to False if the info box content is empty, or True otherwise. I think that being able to explicitly specify that you want to display / hide info boxes is somewhat cleaner.

 - Instead of only building the info window if `display_info_box` is True, I always build it, but only bind the click listener if it's set to be displayed. This makes responding to events somewhat easier, but the idea is basically the same as what you describe.

**Comment by pbugnion** [comment]
I've released v0.3.6, which includes this PR!

**Pull request merged**